### ISSUE TITLE
Add heuristics for add imports quick fixes

### ISSUE THREAD

**Issue by luabud** [microsoft/pylance-release#5226]
Discussed offline with the broader team and we got to an agreement with only showing at most the top 3 high confidence add import options in the quick fix list, and then show an additional source action to "Search for an import match" that opens up the searchable quick pick menu with more import matches.

"High confidence" can be implemented as heuristics around:
- Packages listed in requirements.txt, environment.yml files 
- Packages installed in the environment
- Packages already imported in other files in the workspace

**Comment by Avasam** [comment]
Curious to see what comes out of this. I often get a dozen unrelated autoimport suggestions when working in typeshed

**Comment by heejaechang** [comment]
related issue - https://github.com/microsoft/pyrx/issues/4321

**Comment by luabud** [comment]
@heejaechang I think it makes sense to tackle https://github.com/microsoft/pylance-release/issues/5153 along with this (i.e. to remove the exact match setting). What do you think?

**Comment by heejaechang** [comment]
Here is my preliminary plan for enhancing the `add using` feature.

1. Limit the total number of `add import` suggestions in code action to `3`.
2. Criteria for selecting the 3 are:
  a. Use `MRU` (Most Recently Used) for exact matches (up to 3).
  b. If slots remain, prefer the same module used in other files for the same symbol.
  c. If slots remain, prioritize one from `typing` or `stdlib`, with possible user configuration.
  d. If slots remain, choose from third-party packages.
  e. Otherwise, sort by module and symbol name.
  f. If no exact match exists, include up to 3 fuzzy matched imports.
3. Track around `100` MRUs (100 is an arbitrary number).
4. Introduce a new `search import` in `code action` (marked as refactoring/low priority) that opens a quick pick window, allowing users to search for and `add imports`
   a. e.g., searching `model` to find the correct `django models` when the exact name is forgotten or 
    b. find the correct import name to insert if the original type name differs from the package's re-export. For example, search for `WebDriver` but insert `ChromiumEdge` from `selenium` since `WebDriver` is an internal type name.
    
Goals:

1. Show a limited number of the most probable options for `add imports` in `code actions`.
2. Provide a way for users to find an import even without remembering the precise name of the import/symbol.

**Comment by heejaechang** [comment]
> I think it makes sense to tackle https://github.com/microsoft/pylance-release/issues/5153 along with this (i.e. to remove the exact match setting). What do you think?

sure, we probably will remove it once everything is done. but we probably will do a/b testing and feature rollout and etc? so, existing will stay until we are done with all those?

**Comment by luabud** [comment]
I really like the proposal, thanks for working on that! My one suggestion would be to not do `g. `
i.e. If there aren't exact matches we wouldn't show any import suggestions, only the code action for "Search for an import match".

**Comment by heejaechang** [comment]
> I really like the proposal, thanks for working on that! My one suggestion would be to not do g. 
i.e. If there aren't exact matches we wouldn't show any import suggestions, only the code action for "Search for an import match". That's just my vote though, ofc :)

so we don't even do fuzzy matching for misspelling? currently we have that since `pyright`'s original implementation has it.

**Comment by luabud** [comment]
I was thinking we would in the "show more options" list, but not as top level code actions

**Comment by bschnurr** [comment]
This issue has been fixed in prerelease version 2024.2.100, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20242100-13-February-2024-PreRelease)

**Comment by rchiodo** [comment]
In order to enable this new enhancement, add the following to your settings.json:

```
    "python.analysis.addImport.heuristics": true
```
